### DISCUSSION TITLE
How to write efficient code? ( Micropython)

### DISCUSSION THREAD

**Discussion by carl336** [micropython/micropython#13726] [RP2040 / Pico]
Hi, I am trying to turn on Backlight by using Raspberry Pi Pico with Micropython.
(MicroPython v1.22.0 on 2023-12-27)

Here is the program code that I wrote:

https://paste.fastmirror.net/?c2c2e0dcd1beafe5#F6rycRgyEdiijcoPJoZdrwGWSb9D8BgJJL3PbbPfriav

Someone said that my code looks extremely inefficient, can anyone teach me how to improve my code?

By the way, this is how I set up the LEDs to light up, you can watch it by this URL: 
https://streamable.com/9lx3g3

**Comment by karfas** [comment]
Maybe you should ask *someone* in the first place ?

1) inefficient
```
microPython_code = """
GSCLK = PWM(Pin(8))  # 设定GPIO 20为PWM输出方式
GSCLK.freq(100000)  # 设定频率60HZ
GSCLK.duty_u16(32768)  # 设定占空比，取值范围0~65535
...
"""
```

This is called using `exec(microPython_code)` multiple times from all these `gs_data_IC_x_y()` functions.
Micropython will most likely need to parse/interpret this string every time instead of using an an already parsed function. 
a) Why not make a simple function from that ? 
b) Usually, setting up PWM, Pins, I2C and the like can be done **once**; is there a reason you do this after each key press ?

2) Readability and do-not-repeat-yourself
a) I would create some kind of config file for the different patterns, so these long data definitions no longer live in your code.
b) Maybe some of your long `if ... elsif` cascades can be replaced by arrays or dictionaries, e.g. replace
```python
def LED_pattern_2():
    for i in range(17):
        if i == 0:      # IC No.1
            gs_data_IC_2_1()                  
        elif i == 1:    # IC No.3
            gs_data_IC_2_2()
```
by 
```
# define the data for the pattern somewhere else (config, module).
pattern20 = [ 
    [ 0x05, 0x05, .... ], # data IC_2_1
    [ 0x05, 0x05, .... ], # data IC_2_2
    ...
]

# replace if...elsif by using a simple function call 
gs_data(pattern20[i])
```

**Comment by peterhinch** [comment]
There is [an official doc](https://docs.micropython.org/en/latest/reference/constrained.html) here.

**Comment by scruss** [comment]
Please tell us what hardware you are using. I understand that there may be a language barrier, but if you don't tell us about the hardware, this question will be about as helpful as your previous two identical questions: [Newcomer's request for help, turn on LED](https://github.com/orgs/micropython/discussions/13652) and [Error on Raspberry Pico with Micropython](https://github.com/orgs/micropython/discussions/13627).

Are you using WS2812 LEDs? There's already a [NeoPixel](https://docs.micropython.org/en/v1.22.0/rp2/quickref.html#neopixel-and-apa106-driver) driver built into the RP2040 MicroPython firmware that would make your code much shorter

**Reply by carl336** [reply]
Thanks for your reply.

Firstly, the hardware that I am using is Raspberry Pi Pico with Micropython.
Secondly, I am using KW SITQA1.KD (LEDs).  
Thirdly, I just uploaded some files to a website, you may download them for free to check it.

https://file.io/bThWb9ebu9V4

Those files are:

1. a circuit diagram that I draw,

2. a LED circuit diagram inside the backlight board,

3. a specification for LEDs that are used,

4. PWM LED Driver with Internal Current Settings,

5. A drawing of the specific position of each LED that I draw.

And, the BL that I'm using has 17 LED Drivers, each LED Driver controls 48 LEDs, in total, there are 816 LEDs.

BTW, is it possible to tell me about your social media account?
